SSA/Projects/FullyHomomorphicEncryption/Syntax.lean

```diff
@@ -0,0 +1,127 @@
+
+def mul {Γ : Ctxt (Ty q n)} (e₁ e₂ : Var Γ .polynomialLike) : Expr (Op q n) Γ .polynomialLike :=
+  Expr.mk
+    (op := .mul)
+    (ty_eq := rfl)
+    (args := .cons e₁ <| .cons e₂ .nil)
+    (regArgs := .nil)
+
+def mon {Γ : Ctxt (Ty q n)} (a : Var Γ .integer) (i : Var Γ .index) : Expr (Op q n) Γ .polynomialLike :=
+  Expr.mk
+    (op := .monomial)
+    (ty_eq := rfl)
+    (args := .cons a <| .cons i .nil)
+    (regArgs := .nil)
+
+def mkExpr (Γ : Ctxt (Ty q n)) (opStx : MLIR.AST.Op 0) :
+    MLIR.AST.ReaderM (Op q n) (Σ ty, Expr (Op q n) Γ ty) := do
+  match opStx.name with
+  | "poly.monomial" =>
+    match opStx.args with
+    | v₁Stx::v₂Stx::[] =>
+      let ⟨ty₁, v₁⟩ ← MLIR.AST.TypedSSAVal.mkVal Γ v₁Stx
+      let ⟨ty₂, v₂⟩ ← MLIR.AST.TypedSSAVal.mkVal Γ v₂Stx
+      match ty₁, ty₂ with
+        | .integer, .index => return ⟨.polynomialLike, mon v₁ v₂⟩
+        | _, _ => throw <| .generic s!"expected operands to be of types `integer` and `index` for `monomial`. Got: {repr ty₁}, {repr ty₂}"
+    | _ => throw <| .generic s!"expected two operands for `monomial`, found #'{opStx.args.length}' in '{repr opStx.args}'"
+  | "poly.add" =>
+    match opStx.args with
+    | v₁Stx::v₂Stx::[] =>
+      let ⟨ty₁, v₁⟩ ← MLIR.AST.TypedSSAVal.mkVal Γ v₁Stx
+      let ⟨ty₂, v₂⟩ ← MLIR.AST.TypedSSAVal.mkVal Γ v₂Stx
+      match ty₁, ty₂ with
+        | .polynomialLike, .polynomialLike => return ⟨.polynomialLike, add v₁ v₂⟩
+        | _, _ => throw <| .generic s!"expected both operands to be of type 'polynomialLike'"
+    | _ => throw <| .generic s!"expected two operands for `add`, found #'{opStx.args.length}' in '{repr opStx.args}'"
+  | _ => throw <| .unsupportedOp s!"unsupported operation {repr opStx}"
+
+instance : MLIR.AST.TransformExpr (Op q n) (Ty q n) 0 where
+  mkExpr := mkExpr
+
+def mkReturn (Γ : Ctxt (Ty q n)) (opStx : MLIR.AST.Op 0) : MLIR.AST.ReaderM (Op q n) (Σ ty, Com (Op q n) Γ ty) :=
+  if opStx.name == "return"
+  then match opStx.args with
+  | vStx::[] => do
+    let ⟨ty, v⟩ ← MLIR.AST.TypedSSAVal.mkVal Γ vStx
+    return ⟨ty, Com.ret v⟩
+  | _ => throw <| .generic s!"Ill-formed return statement (wrong arity, expected 1, got {opStx.args.length})"
+  else throw <| .generic s!"Tried to build return out of non-return statement {opStx.name}"
+
+instance : MLIR.AST.TransformReturn (Op q n) (Ty q n) 0 where
+  mkReturn := mkReturn
+
+def mlir2fhe (reg : MLIR.AST.Region 0) :
+    MLIR.AST.ExceptM (Op q n) (Σ (Γ : Ctxt (Ty q n)) (ty : (Ty q n)), Com (Op q n) Γ ty) := MLIR.AST.mkCom reg
+
+end MkFuns -- we don't want q and i here anymore
+
+open Qq MLIR AST Lean Elab Term Meta in
+/--
+NOTE: Wanting Stuck Terms During Strong Normalization of Com
+
+See that we unfold with `TransparencyMode.all`. We do this so we can expose Com.denote
+fully. This is a crazy hack, and interacts extremely poorly with complex definitions found in FHE.
+
+One neater solution is a `match goal` like tactic to match on the proof state
+and run the correct equation.
```

Comment:
In this case, `simp-proc`s should suffice, since we don't need to match on the context. Or really, just proper `Expr` matching, since we don't have to arbitrarily search for sub-expressions.